server/debugsession.go

```diff
@@ -0,0 +1,578 @@
+
+		for _, requestBreakpoint := range requestBreakpoints {
+			ds.debugger.AddBreakpoint(location, uint(requestBreakpoint.Line))
+
+			responseBreakpoints = append(
+				responseBreakpoints,
+				dap.Breakpoint{
+					Source: dap.Source{
+						Path: request.Arguments.Source.Path,
+					},
+					// TODO:
+					Verified: true,
+				},
+			)
+		}
+
+		ds.send(&dap.SetBreakpointsResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.SetBreakpointsResponseBody{
+				Breakpoints: responseBreakpoints,
+			},
+		})
+
+	case *dap.LaunchRequest:
+		// TODO: only allow one program at a time
+
+		var args map[string]any
+		_ = json.Unmarshal(request.Arguments, &args)
+
+		programArg, ok := args["program"]
+		if !ok {
+			ds.send(newDAPErrorResponse(
+				request.Seq,
+				request.Command,
+				dap.ErrorMessage{
+					Format:   "Missing program",
+					ShowUser: true,
+				},
+			))
+			break
+		}
+		ds.code = programArg.(string)
+
+		stopOnEntryArg := args["stopOnEntry"]
+		ds.stopOnEntry, _ = stopOnEntryArg.(bool)
+
+		scriptID := emulator.ComputeScriptID([]byte(ds.code))
+		ds.scriptLocation = common.ScriptLocation(scriptID)
+
+		ds.launchRequested = true
+
+		ds.send(&dap.LaunchResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+		if ds.configurationDone && !ds.launched {
+			ds.run()
+		}
+
+	case *dap.ConfigurationDoneRequest:
+		ds.configurationDone = true
+
+		if ds.launchRequested && !ds.launched {
+			ds.run()
+		}
+
+		ds.send(&dap.ConfigurationDoneResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.ThreadsRequest:
+		ds.send(&dap.ThreadsResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.ThreadsResponseBody{
+				Threads: []dap.Thread{
+					{
+						Id:   1,
+						Name: "Emulator",
+					},
+				},
+			},
+		})
+
+	case *dap.PauseRequest:
+		ds.debugger.RequestPause()
+
+		ds.send(&dap.PauseResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.NextRequest:
+		ds.step()
+
+		ds.send(&dap.NextResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.StepInRequest:
+		// TODO: handled as step request for now
+		ds.step()
+
+		ds.send(&dap.StepInResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.StepOutRequest:
+		// TODO: handled as step request for now
+		ds.step()
+
+		ds.send(&dap.StepOutResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+		})
+
+	case *dap.StackTraceRequest:
+		// TODO: reply with error if ds.stop == nil
+
+		stackFrames := ds.stackFrames()
+
+		ds.send(&dap.StackTraceResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.StackTraceResponseBody{
+				StackFrames: stackFrames,
+			},
+		})
+
+	case *dap.SourceRequest:
+		path := request.Arguments.Source.Path
+
+		code := ds.pathCode(path)
+
```

Comment:
I think here can be better to get codes from CadenceDebugger. So we can support also `Attach`.

Reply:
Agreed, it would be nice to request this from Cadence, but the Cadence debugger and interpreter do not have the source code (only the AST).